### ISSUE TITLE
Circuit breaker not handling exception at end points even when ExceptionsAllowedBeforeBreaking is set.

### ISSUE THREAD

**Issue by anuviswan** [ThreeMammals/Ocelot#1550]
## Expected Behavior / New Feature

As per the documentation of QoS functionality of Ocelot, it uses Polly to achieve Circuit Breaker among the other resilience features. From what I could understand, the following configuration should have triggered the circuit breaker after an endpoint raised exception more than 2 times.

```json
"QoSOptions": {
   "ExceptionsAllowedBeforeBreaking": 2,
   "DurationOfBreak": 10000,
   "TimeoutValue": 2000
}
```

## Actual Behavior / Motivation for New Feature

However, it looked like the exception raised by the end points isn't handled by the circuit breaker. Please note that the Timeout works perfectly and triggers the circuit breaker as expected. However the case wasn't same with the _ExceptionsAllowedBeforeBreaking_.

## Steps to Reproduce the Problem

  1. Create a route in the gateway with the above QoS configuration
  2. The specified Endpoint should raise an exception 
  3. Raise the request for end point via the gateway.

The code sample in the following link demonstrates the issue
[Demo Code](https://github.com/anuviswan/LearningPoint/tree/master/Microservices/AuthenticationDemo)

Similar problem is mentioned in the following Stackoverflow question too
- [Ocelot QoS doesn't break circuit when n number of exception arises](https://stackoverflow.com/questions/56851572/ocelot-qos-doesnt-break-circuit-when-n-number-of-exception-arises)

## Specifications
  - Ocelot: 17.0.1
  - Ocelot.Provider.Polly: 17.0.1

**Comment by ks1990cn** [comment]
Hi,

Thanks for reporting this bug @anuviswan. 
@anuviswan Do you want to propose any solution or anything else you want to add on?

I am able to replicate and started looking into it. @raman-m

**Comment by ks1990cn** [comment]
Raised bug to Polly, https://github.com/App-vNext/Polly/issues/1716.

We are in touch there. Thanks

@anuviswan

**Comment by raman-m** [comment]
@ks1990cn  Tanmay,
Are you sure it's a bug of Polly when posting on Polly boards?
I'm not sure... and the bug can belong to Ocelot, because of ugly Exception-to-Error design, see the [HttpExceptionToErrorMapper](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20HttpExceptionToErrorMapper&type=code) class.

@ggnaegi  Welcome! To have a fun! 😉

**Comment by ks1990cn** [comment]
@raman-m I've tried my own projects and created HttpClient call. After multiple errors from downstream I was unable to find circuitbreaking is working even in my case.
Its not about exception code, exception code is not changing at all in any way. You can see my GIF attached to Polly board.
Ocelot code looks fine to me.

Anyone else also can verify, I will retro inspect my things again.

**Comment by ks1990cn** [comment]
> @ks1990cn Tanmay, Are you sure it's a bug of Polly when posting on Polly boards? I'm not sure... and the bug can belong to Ocelot, because of ugly Exception-to-Error design, see the [HttpExceptionToErrorMapper](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20HttpExceptionToErrorMapper&type=code) class.
> 
> @ggnaegi Welcome! To have a fun! 😉

We don't get code to mapper part even, here is my debugger point
![image](https://github.com/ThreeMammals/Ocelot/assets/29522704/43387f5a-1840-42da-8517-9e738fe844fa)

**Comment by ggnaegi** [comment]
@raman-m @ks1990cn @anuviswan  So, i had a look, I think, the problem is more related with the fact that we are creating a new CircuitBreaker per request.  I have created a new acceptance test for this purpose:

```C#
        [Fact]
        public void should_open_circuit_breaker_after_two_exceptions()
        {
            var port = RandomPortFinder.GetRandomPort();

            var configuration = new FileConfiguration
            {
                Routes = new List<FileRoute>
                {
                    new()
                    {
                        DownstreamPathTemplate = "/",
                        DownstreamScheme = "http",
                        DownstreamHostAndPorts = new List<FileHostAndPort>
                        {
                            new()
                            {
                                Host = "localhost",
                                Port = port,
                            },
                        },
                        UpstreamPathTemplate = "/",
                        UpstreamHttpMethod = new List<string> { "Get" },
                        QoSOptions = new FileQoSOptions
                        {
                            ExceptionsAllowedBeforeBreaking = 2,
                        },
                    },
                },
            };

            this.Given(x => x.GivenThereIsABrokenServiceRunningOn($"http://localhost:{port}"))
                .And(x => _steps.GivenThereIsAConfiguration(configuration))
                .And(x => _steps.GivenOcelotIsRunningWithPolly())
                .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
                .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
                .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
                .Then(x => _steps.ThenTheStatusCodeShouldBe(HttpStatusCode.ServiceUnavailable))
                .BDDfy();
        }
```

**Comment by ks1990cn** [comment]
> @raman-m @ks1990cn @anuviswan Guys, I thought I was writing you rubbish. But, we have a bigger problem here: First, as I wrote before, we shouldn't create a new CircuitBreaker per request / route. It will never work. I was suddenly unsure because in another project we had retry policies aswell.
> 
> I was able to make this acceptance test pass:
> 
> ```cs
> [Fact]
>         public void should_open_circuit_breaker_after_two_exceptions()
>         {
>             var port = RandomPortFinder.GetRandomPort();
> 
>             var configuration = new FileConfiguration
>             {
>                 Routes = new List<FileRoute>
>                 {
>                     new()
>                     {
>                         DownstreamPathTemplate = "/",
>                         DownstreamScheme = "http",
>                         DownstreamHostAndPorts = new List<FileHostAndPort>
>                         {
>                             new()
>                             {
>                                 Host = "localhost",
>                                 Port = port,
>                             },
>                         },
>                         UpstreamPathTemplate = "/",
>                         UpstreamHttpMethod = new List<string> { "Get" },
>                         QoSOptions = new FileQoSOptions
>                         {
>                             ExceptionsAllowedBeforeBreaking = 2,
>                             TimeoutValue = 5000,
>                             DurationOfBreak = 1000,
>                         },
>                     },
>                 },
>             };
> 
>             this.Given(x => x.GivenThereIsABrokenServiceRunningOn($"http://localhost:{port}"))
>                 .And(x => _steps.GivenThereIsAConfiguration(configuration))
>                 .And(x => _steps.GivenOcelotIsRunningWithPolly())
>                 .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
>                 .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
>                 .And(x => _steps.WhenIGetUrlOnTheApiGateway("/"))
>                 .Then(x => _steps.ThenTheStatusCodeShouldBe(HttpStatusCode.ServiceUnavailable))
>                 .BDDfy();
>         }
> ```
> 
> But the code is quite ugly...
> 
> Second, in `PollyCircuitBreakingDelegatingHandler`, no exceptions are thrown here when the client return status 500. I was able to make it work with this terrible bit of code:
> 
> ```cs
> var policyResult = await policy.ExecuteAsync(async () =>
>             {
>                 var result = await base.SendAsync(request, cancellationToken);
>                 if (result.StatusCode == HttpStatusCode.InternalServerError)
>                 {
>                     throw new HttpRequestException("Service Unavailable");
>                 }
> 
>                 return result;
>             });
> 
>             return policyResult;
> ```
> 
> Then, after two tries, the Circuit Breaker kicks in, otherwise never...
> 
> Summary:
> 
> * We shouldn't create a Circuit Breaker per route / request, only maintain one Circuit Breaker per route
> * The Circuit Breaker is Expecting an Exception to kick in,  a response with status 500 is not an Exception
> 
> @ks1990cn is right when he says that the circuit breaker is not working, but it's not a bug in polly, but a bug in ocelot.

Correct I agree on this, I made configuration issue too on my personal project. https://www.youtube.com/watch?v=z6YcU0PW_9E&t=1s helped me to configure it properly. I was doing same mistake. Closed bug on Polly.

**Comment by raman-m** [comment]
The issue has been [accepted](#event-10765116824) due to open PR #1753 
- #1753

**Comment by raman-m** [comment]
> @ggnaegi commented [on Oct 23](#issuecomment-1775459780)
>- But, we have a bigger problem here:
>- the circuit breaker is not working, but it's not a bug in Polly, but a bug in ocelot.

Well... Has this problem already addressed in PR #1753, so is it included in PR?
Sorry, I ask strange questions sometimes

**Comment by ggnaegi** [comment]
> > @ggnaegi commented [on Oct 23](#issuecomment-1775459780)
> >- But, we have a bigger problem here:
> >- the circuit breaker is not working, but it's not a bug in Polly, but a bug in ocelot.
> 
> Well... Has this problem already addressed in PR #1753, so is it included in PR?
> Sorry, I ask strange questions sometimes 
> 

Yes it's part of the PR

**Comment by raman-m** [comment]
@anuviswan  Hey Anu,
Could you check & test the bug fix we've prepared as a gift for you, plz? 😉
